Pytorch Tutorial 5 - MNIST classification - Jupyter and Markdown (#91)
Adamage/tutorials: simple_applications/pytorch/mnist/README.md

          (pool): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
          (relu): ReLU()
        )
        (layer3): Linear(in_features=1600, out_features=128, bias=True)
        (layer3_act): ReLU()
        (layer3_dropout): Dropout(p=0.5, inplace=False)
        (layer4): Linear(in_features=128, out_features=10, bias=True)
        (softmax): Softmax(dim=1)
      )
      (loss): CrossEntropyLoss()
    )

Now we apply the model wrapping function, which will perform a shallow copy
of the PyTorch model. To train the model, we also will use the Stochastic 
Gradient Descent with no momentum [SGD](https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/reference.html#poptorch.optim.SGD).

```python
training_model = poptorch.trainingModel(
    model_with_loss,
    model_opts,
    optimizer=optim.SGD(model.parameters(), lr=learning_rate)
)
```

We are ready to start training. However to track the accuracy while training
we need to define one more helper function. During the training, not every 
samples prediction is returned for efficiency reasons, so this helper function
will check accuracy for labels where prediction is available.

```python

> def accuracy(predictions, labels):

Reviewer: This doesn't seem to do anything useful during training...kind of sits at 100% and jumps to ~85 and back constantly. I would expect to see this slowly increasing during training.

Author: Well yes but it seems the model learns really quickly. Considering this loop operates on `batch_size=8` and `device_iterations=50` each batch has 400 samples. But actually the values we have on the "surface" (as opposed to the IPU internal loop) are just the 8 samples. So if we have 7 predictions correct, we will get 87.5% accuracy, or 100% for 10, or 75% for 6, you see the pattern.
The same happens for the loss values, they still jump every now and then but quickly fall from >2 to close to 0 for majority.
![image](https://user-images.githubusercontent.com/11834161/137717259-ab6a1363-8942-4f67-ab5c-895de4a8ecc9.png)

@dirkbrink  Please let me know if this is enough for clarification, I could also keep the plots in the notebook if you think that it provides some educational value.

Author: @dirkbrink  These are the same plots after setting `AnchorMode.All` . As suspected, the chaotic behavior was due to sampling just the last batch for accuracy and loss prompts for the user.
![image](https://user-images.githubusercontent.com/11834161/137867070-37c577ab-b6c4-416a-b9ff-2c5e395625bc.png)
I will add the Anchor mode explanation and incorporate it permanently into this tutorial.

    _, ind = torch.max(predictions, 1)
    labels = labels[-predictions.size()[0]:]
    accuracy = \
        torch.sum(torch.eq(ind, labels)).item() / labels.size()[0] * 100.0
    return accuracy
```

This code will perform the requested amount of epochs and batches using the
configured Graphcore IPUs.

```python
nr_batches = len(training_data)

for epoch in tqdm(range(1, epochs+1), leave=True, desc="Epochs", total=epochs):
    with tqdm(training_data, total=nr_batches, leave=False) as bar:
        for data, labels in bar:

>             preds, losses = training_model(data, labels)

Reviewer: Same with loss that just fluctuates around 0%...is this actually being calculated correctly?

            mean_loss = torch.mean(losses).item()

            acc = accuracy(preds, labels)
            bar.set_description(
                "Loss: {:0.4f} | Accuracy: {:05.2F}% ".format(mean_loss, acc)
            )
```

Release resources:

```python
training_model.detachFromDevice()
```

Let's check the validation loss on IPU using the trained model. The weights 

> in `model.parameters()` will be copied from the IPU to the host. The trained

Reviewer: The weights from the trained model will be reused...

# PyTorch(PopTorch) MNIST Training Demo

This example demonstrates how to train a network on the MNIST dataset using
PopTorch. To learn more about PopTorch, see our [PyTorch for the IPU: User Guide](https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/index.html).

## How to use this demo

### Environment preparation

Install the Poplar SDK following the instructions in the [Getting Started](https://docs.graphcore.ai/en/latest/getting-started.html)
guide for your IPU system. Make sure to run the `enable.sh` scripts for Poplar 
and PopART and activate a Python3 virtualenv with PopTorch installed.

Then install the package requirements:
```bash
pip install -r requirements.txt
```

### Setting hyperparameters
Set the hyperparameters for this demo. If you're running this example in 
a Jupyter notebook and wish to modify them, re-run all the cells below.

> For further reading on hyperparameters, see [Hyperparameters (machine learning)](https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning))

Reviewer: Based on Louise's comments below, let's get rid of this sentence - our users will know what hyperparameters are

Author: Done in https://github.com/Adamage/tutorials/pull/91/commits/8d8fc4bafc5615c063da28c24f3721f08c12ac1e

# PyTorch(PopTorch) MNIST Training Demo

This example demonstrates how to train a network on the MNIST dataset using
PopTorch.

## How to use this demo

1) Prepare the environment.

Install the Poplar SDK following the instructions in the Getting Started guide 
for your IPU system. Make sure to run the `enable.sh` scripts for Poplar and 
PopART and activate a Python virtualenv with PopTorch installed.

Then install the package requirements:
```bash
pip install -r requirements.txt
```

2) Run the program. Note that the PopTorch Python API only supports Python 3.
Data will be automatically downloaded using torchvision utils.

```bash
python3 mnist_poptorch.py
```

Select your hyperparameters in this cell. If you wish to modify them, re-run
all cells below it. Further reading [Hyperparameters](https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning))
Setup parameters for training:

```python
# Batch size for training
batch_size = 8

# Device iteration - batches per step

> batches_per_step = 50

Reviewer 2: `device_iterations` instead of `batches_per_step`, for consistency with the other resources. It needs to be replaced in other places too.
```suggestion
device_iterations = 50
```

# Batch size for testing
test_batch_size = 80

# Number of epochs to train
epochs = 10

# Learning rate
learning_rate = 0.05
```

Import required libraries:

```python
from tqdm.auto import tqdm
import torch
import torch.nn as nn
import torchvision
import poptorch
import torch.optim as optim
```

Download the datasets for MNIST - database for handwritten digits.
Source: [The MNIST Database](http://yann.lecun.com/exdb/mnist/)

```python

> # The following is a workaround for pytorch issue #1938

Reviewer 2: I don't understand this comment. This issue pytorch/pytorch#1938 seems totally unrelated and is closed. I suppose the actual issue is pytorch/vision#1938 which is also closed.

Author: Yes no longer needed

from six.moves import urllib
opener = urllib.request.build_opener()
opener.addheaders = [("User-agent", "Mozilla/5.0")]

> urllib.request.install_opener(opener)

Reviewer 2: If the original issue has been solved in recent versions of `torchvision`, is this still necessary?

local_dataset_path = 'mnist_data/'

transform_mnist = torchvision.transforms.Compose(
    [
        torchvision.transforms.ToTensor(),
        torchvision.transforms.Normalize((0.1307, ), (0.3081, ))
    ]
)

training_dataset = torchvision.datasets.MNIST(
        local_dataset_path,
        train=True,
        download=True,
        transform=transform_mnist
)

training_data = torch.utils.data.DataLoader(
    training_dataset,
    batch_size=batch_size * batches_per_step,
    shuffle=True,
    drop_last=True
)

test_dataset = torchvision.datasets.MNIST(
        local_dataset_path,
        train=False,
        download=True,
        transform=transform_mnist
)

test_data = torch.utils.data.DataLoader(
    test_dataset,
    batch_size=test_batch_size,
    shuffle=True,
    drop_last=True
)
```

> Let's define the elements of our neural network. First, we create the class

Reviewer 3: Readers are assumed to be ML engineers and knowledgeable about Pytorch, so we can keep Pytorch-specific explanations short: "Let's define the elements of our neural network. We first create a `Block` instance consisting of a 2D convolutional layer with pooling, followed by a ReLU activation".
We can drop links to Wikipedia too and assume that readers know about ML concepts.

`Block` which will consist of a simple 2D convolutional layer with pooling and
a rectified linear unit (ReLU). To see explanation of pooling and ReLU, see:
[Convolutional Neural Network](https://en.wikipedia.org/wiki/Convolutional_neural_network#Building_blocks)

```python
class Block(nn.Module):
    def __init__(self, in_channels, num_filters, kernel_size, pool_size):
        super(Block, self).__init__()
        self.conv = nn.Conv2d(in_channels,
                              num_filters,
                              kernel_size=kernel_size)
        self.pool = nn.MaxPool2d(kernel_size=pool_size)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = self.conv(x)
        x = self.pool(x)
        x = self.relu(x)
        return x
```

> Now let's construct the deep neural network with 4 Convolutional layers and

Reviewer 3: It only has 2 convolution layers - layers 3 and 4 are linear layers (which is why in `forward`, the output of layer 2 needs to be flattened so it can be passed to the next Linear layer).
You can turn this into "Now, let's construct our neural network." without describing the content (it's a short network so the code is easy to read)

a [softmax layer](https://en.wikipedia.org/wiki/Softmax_function) at the output.

```python
class Network(nn.Module):
    def __init__(self):
        super(Network, self).__init__()
        self.layer1 = Block(1, 32, 3, 2)
        self.layer2 = Block(32, 64, 3, 2)
        self.layer3 = nn.Linear(1600, 128)
        self.layer3_act = nn.ReLU()
        self.layer3_dropout = torch.nn.Dropout(0.5)
        self.layer4 = nn.Linear(128, 10)

>         self.softmax = nn.Softmax(1)

Reviewer 2: There's a `softmax` here, and then we use later a `CrossEntropyLoss`, which already uses a softmax (`LogSoftmax` more precisely). Can we either remove this `softmax`, or replace the `CrossEntropyLoss` with an `NLLLoss`?

Author: Let me check

    def forward(self, x):
        x = self.layer1(x)
        x = self.layer2(x)
        # Flatten layer
        x = x.view(-1, 1600)
        x = self.layer3_act(self.layer3(x))
        x = self.layer4(self.layer3_dropout(x))
        x = self.softmax(x)
        return x
```

Here we define a thin wrapper around the `torch.nn.Module` that will use
cross-entropy loss function - see more [here](https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_loss_function_and_logistic_regression)

This class is creating a custom module to compose the Neural Network and 
the Cross Entropy module into one object, which under the hood will invoke 
the `__call__` function on `nn.Module` and consequently the `forward` method 
when called like this:
```python
prediction, losses = TrainingModelWithLoss(Network())(data, labels)
```

```python
class TrainingModelWithLoss(torch.nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
        self.loss = torch.nn.CrossEntropyLoss()

    def forward(self, args, loss_inputs=None):
        output = self.model(args)
        loss = self.loss(output, loss_inputs)
        return output, loss
```

> Let's initiate the neural network from our defined classes.

Author: Yes :)

```python
model = Network()
model_with_loss = TrainingModelWithLoss(model)
model_opts = poptorch.Options().deviceIterations(batches_per_step)
```

We can check if the model is assembled correctly by printing the string 
representation of the model object

```python
print(model_with_loss)
```

    TrainingModelWithLoss(
      (model): Network(
        (layer1): Block(
          (conv): Conv2d(1, 32, kernel_size=(3, 3), stride=(1, 1))
          (pool): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
          (relu): ReLU()
        )
        (layer2): Block(
          (conv): Conv2d(32, 64, kernel_size=(3, 3), stride=(1, 1))
          (pool): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
          (relu): ReLU()
        )
        (layer3): Linear(in_features=1600, out_features=128, bias=True)
        (layer3_act): ReLU()
        (layer3_dropout): Dropout(p=0.5, inplace=False)
        (layer4): Linear(in_features=128, out_features=10, bias=True)
        (softmax): Softmax(dim=1)
      )
      (loss): CrossEntropyLoss()
    )

Now we apply the model wrapping function, which will perform a shallow copy

> of the PyTorch model. To perform the machine learning operations, we also

Reviewer 2: ```suggestion
of the PyTorch model. To train the model, we also
```

Reviewer 2: "machine learning operations" doesn't tell what we're doing next here

```python
training_model = poptorch.trainingModel(
    model_with_loss,
    model_opts,
    optimizer=optim.SGD(model.parameters(), lr=learning_rate)
)
```

We are ready to start training. However to track the accuracy while training
we need to define one more helper function. During the training, not every 
samples prediction is returned for efficiency reasons, so this helper function
will check accuracy for labels where prediction is available.

```python
def accuracy(predictions, labels):
    _, ind = torch.max(predictions, 1)
    labels = labels[-predictions.size()[0]:]
    accuracy = \
        torch.sum(torch.eq(ind, labels)).item() / labels.size()[0] * 100.0
    return accuracy
```

This code will perform the requested amount of epochs and batches using the
configured Graphcore IPUs.

```python
nr_batches = len(training_data)

for epoch in tqdm(range(1, epochs+1), leave=True, desc="Epochs", total=epochs):
    with tqdm(training_data, total=nr_batches, leave=False) as bar:
        for data, labels in bar:
            preds, losses = training_model(data, labels)

>             with torch.inference_mode():

Reviewer 2: This is not needed as autograd is not enabled by this loop, and doesn't influence the PopTorch model
```suggestion

```

                mean_loss = torch.mean(losses).item()
                acc = accuracy(preds, labels)
            bar.set_description(
                "Loss: {:0.4f} | Accuracy: {:05.2F}% ".format(mean_loss, acc)
            )
```

> Update the weights in model by copying from the training IPU. 
> This updates `model.parameters()`.
> 
> 
> ```python
> training_model.copyWeightsToHost()
> ```

Reviewer 2: This is done implicitly now. We do not need it.
```suggestion

```

Release resources:

```python
training_model.detachFromDevice()
```

Check validation loss on IPU once trained. Because PopTorch will be compiled 
on first call the weights in `model.parameters()` will be copied implicitly. 

> Subsequent calls will need to call `inference_model.copyWeightsToDevice()`.

Reviewer 2: PopTorch is not compiled on the first call, but the model is. Also, we do not need to call `inference_model.copyWeightsToDevice()`. This is really confusing here.
Instead, by reusing the [trained] model, the weights will be copied from the IPU back to the host, and they will be used to compile a new inference model.

Author: Changed the description please check

```python
inference_model = poptorch.inferenceModel(model)
```

Perform validation

```python
nr_batches = len(test_data)
sum_acc = 0.0

> with torch.no_grad():

Reviewer 2: Again, no needed, there's no autograd involved.
```suggestion

```

# PyTorch(PopTorch) MNIST Training Demo

This example demonstrates how to train a network on the MNIST dataset using
PopTorch. To learn more about PopTorch, see our [PyTorch for the IPU: User Guide](https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/index.html).

> ## How to use this demo

Reviewer: I feel like the 1,2,3 sections under this heading don't really make sense in the context of a notebook. Number 2 only matters when running the script itself. So please reword it. Also the hyperparameter setting in section 3 is fine, but the ordering doesn't make sense as it is now. Also, there is no new heading for the whole rest of the tutorial, so technically everything is under 3...probably needs a new heading to split up the rest of the notebook from the "How to use this demo" section

    def __init__(self, in_channels, num_filters, kernel_size, pool_size):
        super(Block, self).__init__()
        self.conv = nn.Conv2d(in_channels,
                              num_filters,
                              kernel_size=kernel_size)
        self.pool = nn.MaxPool2d(kernel_size=pool_size)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = self.conv(x)
        x = self.pool(x)
        x = self.relu(x)
        return x
```

Now, let's construct our neural network.

```python
class Network(nn.Module):
    def __init__(self):
        super(Network, self).__init__()
        self.layer1 = Block(1, 32, 3, 2)
        self.layer2 = Block(32, 64, 3, 2)
        self.layer3 = nn.Linear(1600, 128)
        self.layer3_act = nn.ReLU()
        self.layer3_dropout = torch.nn.Dropout(0.5)
        self.layer4 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.layer1(x)
        x = self.layer2(x)
        # Flatten layer
        x = x.view(-1, 1600)
        x = self.layer3_act(self.layer3(x))
        x = self.layer4(self.layer3_dropout(x))
        return x
```

Next we define a thin wrapper around the `torch.nn.Module` that will use

> the cross-entropy loss function. To learn more about cross entropy click [here](https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_loss_function_and_logistic_regression).

Reviewer 3: You can remove "To learn more about [...]" (same reasoning as above, intended audience are ML engineers who should know about this)

the Cross Entropy module into one object, which under the hood will invoke 
the `__call__` function on `nn.Module` and consequently the `forward` method 
when called like this:
```python
prediction, losses = TrainingModelWithLoss(Network())(data, labels)
```

```python
class TrainingModelWithLoss(torch.nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
        self.loss = torch.nn.CrossEntropyLoss()

    def forward(self, args, loss_inputs=None):
        output = self.model(args)
        loss = self.loss(output, loss_inputs)
        return output, loss
```

Let's initialise the neural network from our defined classes.

```python
model = Network()
model_with_loss = TrainingModelWithLoss(model)
model_opts = poptorch.Options().deviceIterations(device_iterations)
```

Next we will set the `AnchorMode` for our training. By default, `poptorch` will
return to the host machine only a limited set of information for performance
reasons. By default, only the last batch of the internal loop is returned which
is represented by setting `AnchorMode.Final`. When inspecting the training

> performance as it is executing, values like accuracy or losses value will be

Reviewer 3: ```suggestion
performance as it is executing, values like accuracy or losses will be
```

        self.layer3_act = nn.ReLU()
        self.layer3_dropout = torch.nn.Dropout(0.5)
        self.layer4 = nn.Linear(128, 10)

>         self.softmax = nn.Softmax(1)

Reviewer: Why did you remove the softmax?

Author: It was suggested by other reviewer. It's because cross entropy already includes this calculation. If I were to keep it, we could change cross entropy loss to NLL loss
